File: python/trace.py
from __future__ import annotations

import functools
from contextlib import contextmanager
from dataclasses import dataclass, field
from typing import Any, Callable, Iterator, TypeVar

import otel

F = TypeVar("F", bound=Callable[..., Any])

TS_FIELDS: tuple[str, ...] = ("ts_ms", "created_at", "audio_start_ms", "audio_end_ms")
FLOAT_FIELDS: tuple[str, ...] = (
    "score",
    "eou.score",
    "eou.eager_score",
    "eou.threshold",
    "vad.probability",
)
ID_PREFIXES: tuple[tuple[str, str], ...] = (
    ("sess_", "sess"),
    ("item_", "item"),
    ("resp_", "resp"),
    ("evt_", "evt"),
)

@dataclass
class CanonicalTrace:
    events: list[Any] = field(default_factory=list)

    def __len__(self) -> int:
        return len(self.events)

    def is_empty(self) -> bool:
        return not self.events

def canonicalize_trace(trace: list[Any]) -> CanonicalTrace:
    id_map: dict[str, str] = {}
    counters: dict[str, int] = {}
    out: list[Any] = []
    for ev in trace:
        cloned = _deepcopy_jsonish(ev)
        _canonicalize_node(cloned, id_map, counters, parent_key=None)
        out.append(cloned)
    return CanonicalTrace(events=out)

def _deepcopy_jsonish(v: Any) -> Any:
    if isinstance(v, dict):
        return {k: _deepcopy_jsonish(x) for k, x in v.items()}
    if isinstance(v, list):
        return [_deepcopy_jsonish(x) for x in v]
    return v

def _canonicalize_node(
    v: Any,
    id_map: dict[str, str],
    counters: dict[str, int],
    parent_key: str | None,
) -> Any:
    if isinstance(v, dict):
        for k in list(v.keys()):
            child = v[k]
            if k in TS_FIELDS:
                if isinstance(child, int) and not isinstance(child, bool) and child >= 0:
                    v[k] = 0
                continue
            if k in FLOAT_FIELDS:
                if isinstance(child, (int, float)) and not isinstance(child, bool):
                    v[k] = round(float(child), 3)
                continue
            if isinstance(child, str):
                canon = _canon_id(child, id_map, counters)
                if canon is not None:
                    v[k] = canon
                elif k in ("audio", "data"):
                    v[k] = f"<{len(child)} bytes>"
                continue
            _canonicalize_node(child, id_map, counters, parent_key=k)
        return v
    if isinstance(v, list):
        for i, item in enumerate(v):
            if isinstance(item, str):
                canon = _canon_id(item, id_map, counters)
                if canon is not None:
                    v[i] = canon
            else:
                _canonicalize_node(item, id_map, counters, parent_key=parent_key)
        return v
    if isinstance(v, str):
        canon = _canon_id(v, id_map, counters)
        if canon is not None:
            return canon
    return v

def _canon_id(
    s: str,
    id_map: dict[str, str],
    counters: dict[str, int],
) -> str | None:
    for prefix, kind in ID_PREFIXES:
        if s.startswith(prefix):
            existing = id_map.get(s)
            if existing is not None:
                return existing
            n = counters.get(kind, 0) + 1
            counters[kind] = n
            canon = f"{kind}_{n}"
            id_map[s] = canon
            return canon
    return None

def trace_diff(a: CanonicalTrace, b: CanonicalTrace) -> int | None:
    n = min(len(a.events), len(b.events))
    for i in range(n):
        if a.events[i] != b.events[i]:
            return i
    if len(a.events) != len(b.events):
        return n
    return None

def init() -> bool:
    return otel.init()

@contextmanager
def span(name: str, **attributes: Any) -> Iterator[Any]:
    if not otel.is_enabled():
        yield None
        return
    try:
        from opentelemetry import trace as ot_trace

        tracer = ot_trace.get_tracer(otel.TRACER_NAME)
        with tracer.start_as_current_span(name) as sp:
            for k, val in attributes.items():
                try:
                    sp.set_attribute(k, val)
                except Exception:
                    pass
            yield sp
    except ImportError:
        yield None
    except Exception:
        yield None

def traced(name: str | None = None, **span_attributes: Any) -> Callable[[F], F]:
    def decorator(fn: F) -> F:
        span_name = name if name is not None else f"{fn.__module__}.{fn.__qualname__}"

        @functools.wraps(fn)
        def wrapper(*args: Any, **kwargs: Any) -> Any:
            with span(span_name, **span_attributes):
                return fn(*args, **kwargs)

        return wrapper  # type: ignore[return-value]

    return decorator

File: rust/src/trace.rs
#![allow(dead_code)]

use std::collections::HashMap;

use serde_json::Value;

const TS_FIELDS: &[&str] = &["ts_ms", "created_at", "audio_start_ms", "audio_end_ms"];
const FLOAT_FIELDS: &[&str] = &[
    "score",
    "eou.score",
    "eou.eager_score",
    "eou.threshold",
    "vad.probability",
];

const ID_PREFIXES: &[(&str, &str)] = &[
    ("sess_", "sess"),
    ("item_", "item"),
    ("resp_", "resp"),
    ("evt_", "evt"),
];

#[derive(Clone, Debug)]
pub struct CanonicalTrace {
    pub events: Vec<Value>,
}

impl CanonicalTrace {
    pub fn len(&self) -> usize {
        self.events.len()
    }

    pub fn is_empty(&self) -> bool {
        self.events.is_empty()
    }
}

pub fn canonicalize_trace(trace: &[Value]) -> CanonicalTrace {
    let mut id_map: HashMap<String, String> = HashMap::new();
    let mut counters: HashMap<&'static str, usize> = HashMap::new();

    let mut out = Vec::with_capacity(trace.len());
    for ev in trace.iter() {
        let mut cloned = ev.clone();
        canonicalize_node(&mut cloned, &mut id_map, &mut counters);
        out.push(cloned);
    }
    CanonicalTrace { events: out }
}

fn canonicalize_node(
    v: &mut Value,
    id_map: &mut HashMap<String, String>,
    counters: &mut HashMap<&'static str, usize>,
) {
    match v {
        Value::Object(map) => {
            let keys: Vec<String> = map.keys().cloned().collect();
            for k in keys {
                if let Some(child) = map.get_mut(&k) {
                    if TS_FIELDS.iter().any(|f| *f == k) {
                        if let Some(_n) = child.as_u64() {
                            *child = Value::from(0u64);
                        }
                        continue;
                    }
                    if FLOAT_FIELDS.iter().any(|f| *f == k) {
                        if let Some(f) = child.as_f64() {
                            let r = (f * 1000.0).round() / 1000.0;
                            *child = serde_json::Number::from_f64(r)
                                .map(Value::Number)
                                .unwrap_or(Value::Null);
                        }
                        continue;
                    }
                    if let Value::String(s) = child {
                        if let Some(canon) = canon_id(s, id_map, counters) {
                            *s = canon;
                        }
                        continue;
                    }
                    if k == "audio" || k == "data" {
                        if let Value::String(s) = child {
                            let n = s.len();
                            *child = Value::from(format!("<{n} bytes>"));
                            continue;
                        }
                    }
                    canonicalize_node(child, id_map, counters);
                }
            }
        }
        Value::Array(items) => {
            for it in items.iter_mut() {
                canonicalize_node(it, id_map, counters);
            }
        }
        Value::String(s) => {
            if let Some(canon) = canon_id(s, id_map, counters) {
                *s = canon;
            }
        }
        _ => {}
    }
}

fn canon_id(
    s: &str,
    id_map: &mut HashMap<String, String>,
    counters: &mut HashMap<&'static str, usize>,
) -> Option<String> {
    for (prefix, kind) in ID_PREFIXES {
        if s.starts_with(prefix) {
            if let Some(c) = id_map.get(s) {
                return Some(c.clone());
            }
            let n = counters.entry(kind).or_insert(0);
            *n += 1;
            let canon = format!("{kind}_{n}");
            id_map.insert(s.to_string(), canon.clone());
            return Some(canon);
        }
    }
    None
}

pub fn trace_diff(a: &CanonicalTrace, b: &CanonicalTrace) -> Option<usize> {
    let n = a.events.len().min(b.events.len());
    for i in 0..n {
        if a.events[i] != b.events[i] {
            return Some(i);
        }
    }
    if a.events.len() != b.events.len() {
        return Some(n);
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn canonicalize_replaces_random_ids_with_counters() {
        let trace = vec![
            json!({"type": "session.created", "session": {"id": "sess_abc"}}),
            json!({"type": "input_audio_buffer.speech_started", "item_id": "item_xy"}),
            json!({"type": "input_audio_buffer.speech_stopped", "item_id": "item_xy"}),
            json!({"type": "response.created", "response": {"id": "resp_q"}}),
        ];
        let c = canonicalize_trace(&trace);
        assert_eq!(c.events[0]["session"]["id"], "sess_1");
        assert_eq!(c.events[1]["item_id"], "item_1");
        assert_eq!(c.events[2]["item_id"], "item_1");
        assert_eq!(c.events[3]["response"]["id"], "resp_1");
    }

    #[test]
    fn canonicalize_strips_timestamps() {
        let trace = vec![json!({
            "type": "input_audio_buffer.speech_stopped",
            "audio_end_ms": 12345,
            "ts_ms": 99999
        })];
        let c = canonicalize_trace(&trace);
        assert_eq!(c.events[0]["audio_end_ms"], 0);
        assert_eq!(c.events[0]["ts_ms"], 0);
    }

    #[test]
    fn canonicalize_rounds_floats() {
        let trace = vec![json!({
            "type": "eou.scored",
            "score": 0.123456789_f64,
            "threshold": 0.5_f64
        })];
        let c = canonicalize_trace(&trace);
        let s = c.events[0]["score"].as_f64().unwrap();
        assert!((s - 0.123).abs() < 1e-6, "got {s}");
    }

    #[test]
    fn trace_diff_returns_first_diverging_index() {
        let a = canonicalize_trace(&[
            json!({"type": "session.created", "session": {"id": "sess_1"}}),
            json!({"type": "response.created", "response": {"id": "resp_1"}}),
            json!({"type": "response.done", "response": {"id": "resp_1", "status": "completed"}}),
        ]);
        let b = canonicalize_trace(&[
            json!({"type": "session.created", "session": {"id": "sess_z"}}),
            json!({"type": "response.created", "response": {"id": "resp_z"}}),
            json!({"type": "response.done", "response": {"id": "resp_z", "status": "cancelled"}}),
        ]);
        assert_eq!(trace_diff(&a, &b), Some(2));
    }

    #[test]
    fn trace_diff_returns_none_for_identical() {
        let a = canonicalize_trace(&[json!({"type": "x", "id": "evt_a"})]);
        let b = canonicalize_trace(&[json!({"type": "x", "id": "evt_b"})]);
        assert_eq!(trace_diff(&a, &b), None);
    }
}
